File: 2025/go/day-6/part-2.go
package day6

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	FILE_PATH_P2 = "../inputs/day-6/part-1/input.txt"
)

func readLines(filePath string) ([]string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func padLines(lines []string) []string {
	maxLen := 0
	for _, line := range lines {
		if len(line) > maxLen {
			maxLen = len(line)
		}
	}

	padded := make([]string, len(lines))
	for i, line := range lines {
		if len(line) < maxLen {
			padded[i] = line + strings.Repeat(" ", maxLen-len(line))
		} else {
			padded[i] = line
		}
	}
	return padded
}

func isColumnEmpty(lines []string, col int) bool {
	for row := 0; row < len(lines); row++ {
		if col < len(lines[row]) && lines[row][col] != ' ' {
			return false
		}
	}
	return true
}

func readNumberVertically(lines []string, col, startRow int) (int, int, error) {
	numStr := ""
	row := startRow
	for row < len(lines) {
		if col >= len(lines[row]) {
			break
		}
		ch := lines[row][col]
		if ch >= '0' && ch <= '9' {
			numStr += string(ch)
			row++
		} else {
			break
		}
	}
	if numStr == "" {
		return 0, startRow, fmt.Errorf("no number found")
	}
	num, err := strconv.Atoi(numStr)
	if err != nil {
		return 0, startRow, err
	}
	return num, row - 1, nil
}

func processProblem(lines []string, startCol int) ([]int, string, int) {
	var numbers []int
	var operator string
	col := startCol

	for col >= 0 {
		if isColumnEmpty(lines, col) {
			break
		}

		for row := 0; row < len(lines); row++ {
			if col >= len(lines[row]) {
				continue
			}

			char := lines[row][col]
			if char == ' ' {
				continue
			}

			if char == '+' || char == '*' {
				operator = string(char)
			} else if char >= '0' && char <= '9' {
				num, lastRow, err := readNumberVertically(lines, col, row)
				if err == nil {
					numbers = append(numbers, num)
					row = lastRow
				}
			}
		}
		col--
	}
	return numbers, operator, col
}

func calculateProblem(numbers []int, operator string) int {
	if len(numbers) == 0 || operator == "" {
		return 0
	}

	switch operator {
	case "+":
		return addition(numbers)
	case "*":
		return multiplication(numbers)
	default:
		return 0
	}
}

func Day6Part2() {
	file, err := os.Open(FILE_PATH_P2)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		os.Exit(1)
	}

	if len(lines) == 0 {
		fmt.Println("No lines found")
		return
	}

	lines = padLines(lines)
	grandTotal := 0
	col := len(lines[0]) - 1

	for col >= 0 {
		for col >= 0 && isColumnEmpty(lines, col) {
			col--
		}

		if col < 0 {
			break
		}

		numbers, operator, nextCol := processProblem(lines, col)
		result := calculateProblem(numbers, operator)
		grandTotal += result
		col = nextCol
	}

	fmt.Printf("Solution: %d\n", grandTotal)
}

File: 2025/ts/src/day-6/part-2.ts
import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';

const PATH = path.join(__dirname, '../../..', 'inputs/day-6/part-1/input.txt');

function addition(nums: number[]): number {
  let total = 0;
  for (const num of nums) {
    total += num;
  }
  return total;
}

function multiplication(nums: number[]): number {
  let total = 1;
  for (const num of nums) {
    total *= num;
  }
  return total;
}

async function readLines(filePath: string): Promise<string[]> {
  const fileStream = fs.createReadStream(filePath);
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  });

  const lines: string[] = [];
  for await (const line of rl) {
    lines.push(line);
  }
  return lines;
}

function padLines(lines: string[]): string[] {
  let maxLen = 0;
  for (const line of lines) {
    if (line.length > maxLen) {
      maxLen = line.length;
    }
  }

  const padded: string[] = [];
  for (const line of lines) {
    if (line.length < maxLen) {
      padded.push(line + ' '.repeat(maxLen - line.length));
    } else {
      padded.push(line);
    }
  }
  return padded;
}

function isColumnEmpty(lines: string[], col: number): boolean {
  for (let row = 0; row < lines.length; row++) {
    if (col < lines[row].length && lines[row][col] !== ' ') {
      return false;
    }
  }
  return true;
}

function readNumberVertically(lines: string[], col: number, startRow: number): [number, number, boolean] {
  let numStr = '';
  let row = startRow;
  while (row < lines.length) {
    if (col >= lines[row].length) {
      break;
    }
    const ch = lines[row][col];
    if (ch >= '0' && ch <= '9') {
      numStr += ch;
      row++;
    } else {
      break;
    }
  }

  if (numStr === '') {
    return [0, startRow, false];
  }

  const num = parseInt(numStr, 10);
  if (isNaN(num)) {
    return [0, startRow, false];
  }

  return [num, row - 1, true];
}

function processProblem(lines: string[], startCol: number): [number[], string, number] {
  const numbers: number[] = [];
  let operator = '';
  let col = startCol;

  while (col >= 0) {
    if (isColumnEmpty(lines, col)) {
      break;
    }

    for (let row = 0; row < lines.length; row++) {
      if (col >= lines[row].length) {
        continue;
      }

      const char = lines[row][col];
      if (char === ' ') {
        continue;
      }

      if (char === '+' || char === '*') {
        operator = char;
      } else if (char >= '0' && char <= '9') {
        const [num, lastRow, success] = readNumberVertically(lines, col, row);
        if (success) {
          numbers.push(num);
          row = lastRow;
        }
      }
    }
    col--;
  }

  return [numbers, operator, col];
}

function calculateProblem(numbers: number[], operator: string): number {
  if (numbers.length === 0 || operator === '') {
    return 0;
  }

  switch (operator) {
    case '+':
      return addition(numbers);
    case '*':
      return multiplication(numbers);
    default:
      return 0;
  }
}

export async function day6Part2(): Promise<void> {
  try {
    const lines = await readLines(PATH);

    if (lines.length === 0) {
      console.log('No lines found');
      return;
    }

    const paddedLines = padLines(lines);
    let grandTotal = 0;
    let col = paddedLines[0].length - 1;

    while (col >= 0) {
      while (col >= 0 && isColumnEmpty(paddedLines, col)) {
        col--;
      }

      if (col < 0) {
        break;
      }

      const [numbers, operator, nextCol] = processProblem(paddedLines, col);
      const result = calculateProblem(numbers, operator);
      grandTotal += result;
      col = nextCol;
    }

    console.log(`Solution: ${grandTotal}`);
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}